torch/distributed/_spmd/api.py

```diff
@@ -204,6 +205,87 @@ def _rematerialize_optimizer(
         opt.state.update(orig_states)
 
 
+aten = torch.ops.aten  # pyre-ignore
+
+FOREACH_OP_MAP = {
+    aten._foreach_add_.List: aten._foreach_add.List,
+}
+
+
+def getitem(l, i):
+    return l[i]
+
+
+def _functionalize_and_reinplace_foreach_ops(
+    gm: fx.GraphModule,
+) -> fx.GraphModule:
+    # inplace foreach ops been replaced
+    inplace_updated: Dict[fx.Node, fx.Node] = {}
+    # skip newly inserted nodes
+    skip_nodes: Set[fx.Node] = set()
+
+    for node in gm.graph.nodes:
+        if node in skip_nodes:
+            continue
+
+        flat_args, spec = pytree.tree_flatten(node.args)
+        updated_args = [inplace_updated.get(a) or a for a in flat_args]
+        new_args = pytree.tree_unflatten(updated_args, spec)
+
+        # If we got new arg, it means this node consumes inplace results.
+        # Replace this node with a new one that uses functionalized outputs.
+        if new_args != node.args:
+            with gm.graph.inserting_after(node):
+                new_node = gm.graph.create_node(
+                    node.op,
+                    node.target,
+                    new_args,
+                    node.kwargs,
+                    node.name,
+                    node.type,
+                )
+                node.replace_all_uses_with(new_node)
+            gm.graph.erase_node(node)
+            continue
+
+        # Only functionalize aten._foreach_* ops.
+        if node.op == OP.CALL_FUNCTION and node.target in FOREACH_OP_MAP:
+            with gm.graph.inserting_after(node):
+                # N.B.: inplace foreach does not return anything, so its output
+                # is not connected to any subsequent nodes anyway. It's safe to
+                # create a out of place version and remove the inplace one.
+                new_node = gm.graph.call_function(
+                    FOREACH_OP_MAP[node.target],
+                    args=node.args,
+                    kwargs=node.kwargs,
+                )
+                getitems: Dict[fx.Node, fx.Node] = {}
+                # need another inserting_after, otherwise new ones are always
+                # inserted immediately after ``node``
```

Comment:
yup, this is annoying. Should we just remove the one level of indentation of this `inserting_after` and make it have as the same as the previous `inserting_after`? The previous `inserting_after` is useless for the L265-L271.